runtime/doc/lua-plugin.txt

```diff
@@ -0,0 +1,290 @@
+In this tutorial let's use the following path:
+`$XDG_DATA_HOME/nvim/site/pack/tutorial/start/no-trailing` (see |$XDG_DATA_HOME
+for its value on your system). The reason to choose this complicated path is
+because it utilizes built-in |packages| functionality to make this plugin
+installed in your config.
+
+The newly created `no-trailing/` directory will be treated as root for all
+relative paths used here. In particular as a first commented line in code
+blocks to describe where it should be placed.
+
+Now let's create a `lua/no-trailing/init.lua` file. It will contain the Lua
+code implementing core functionality. For plugin that removes trailing lines
+it can be something like this:
+>lua
+    -- lua/no-trailing/init.lua
+
+    -- Create a table to store and later return module's user facing functions
+    -- Using `M` is a popular convention, but any name is possible
+    local M = {}
+
+    -- Create user facing function
+    M.trim = function(buf_id, opts)
+      -- Validate required arguments
+      local ok, is_buf_valid = pcall(vim.api.nvim_buf_is_valid, buf_id)
+      assert(ok and is_buf_valid, '`buf_id` is not a valid buffer identifier')
+
+      -- Infer and validate optional arguments
+      opts = vim.tbl_deep_extend('force', { empty = false }, opts or {})
+      assert(type(opts.empty) == 'boolean', '`opts.empty` is not boolean')
+
+      -- Compute number of the first line to be removed: iterate from the end
+      -- backwards and stop when line is not target depending on `empty`
+      local target_pattern = opts.empty and '^$' or '^%s*$'
+      local lines = vim.api.nvim_buf_get_lines(buf_id, 0, -1, false)
+      local start_line
+      for i = #lines, 1, -1 do
+        if lines[i]:find(target_pattern) == nil then
+          start_line = i
+          break
+        end
+      end
+      start_line = start_line or 0
+
+      -- Remove trailing lines
+      vim.api.nvim_buf_set_lines(buf_id, start_line, -1, false, {})
+    end
+
+    -- Return the module. This must be the last of the file.
+    return M
+<
+And this is it. You have create a Lua plugin. Users now can install it
+(i.e. put the whole `no-trailing/` directory in their 'runtimepath' manually
+or via plugin manager) and use it directly with `require()`:
+
+• Any of `require('no-trailing').trim(0)`, `require('no-trailing').trim(0, {})`,
+  or `require('no-trailing').trim(0, { empty = false })` removes all
+  blank trailing lines.
+
+• `require('no-trailing').trim(0, { empty = true })` removes only empty
+  trailing lines.
+
+• Any of `require('no-trailing').trim()`, `require('no-trailing').setup('a')`,
+  `require('no-trailing').trim(0, { empty = 'a' })` will throw an error.
+
+Notes:
+
+• Choosing function scope (i.e. what it does) is up to you. Usually preferring
+  the functionality just enough for it to be useful for target audience is
+  a good approach. You can see what Nvim does in |dev-lua|.
```

Comment:
Huh? This didn't help at all. I don't know what the takeaway from this is or what "guiding" is done here.